lib/trento/support/event_handler_failure_context.ex

```diff
@@ -0,0 +1,73 @@
+defmodule Trento.Support.EventHandlerFailureContext do
+  @moduledoc """
+  Event handler failure context
+
+  max_retries: max retries before the event handler is shut down (defauklt: 3)
+  retry_after: time between retries in ms (default: 0)
+  after_retry: callback to be called after reach retry
+  after_max_retries_reached: callback to be called when the max retries are reached
+  skip: if skip is true, the event will be skipped, otherwise the process will stop (default: false)
+  """
+
+  defmacro __using__(opts \\ []) do
+    quote do
+      alias Commanded.Event.FailureContext
+
+      require Logger
+
+      def error(
+            error,
+            event,
+            %FailureContext{context: context, metadata: metadata} = failure_context
+          ) do
+        max_retries = Keyword.get(unquote(opts), :max_retries, 3)
+        retry_after = Keyword.get(unquote(opts), :retry_after, 500)
+        skip = Keyword.get(unquote(opts), :skip, false)
+
+        after_max_retries_reached =
+          Keyword.get(unquote(opts), :after_max_retries_reached, fn _event, _metadata, _context ->
+            :ok
+          end)
+
+        after_retry =
+          Keyword.get(unquote(opts), :after_retry, fn _event, _metadata, _context ->
+            :ok
+          end)
+
+        Logger.metadata(error: error, event: event, failure_context: failure_context)
+
+        case record_failure(context) do
+          %{failures: failures} when failures >= max_retries ->
+            :ok = after_max_retries_reached.(event, metadata, context)
+
+            if skip do
+              Logger.error(
+                "#{__MODULE__} failed to handle event, skipping after #{max_retries} retries"
+              )
+
+              :skip
+            else
+              Logger.error(
+                "#{__MODULE__} failed to handle event, stopping after #{max_retries} retries"
+              )
+
+              {:error, :max_retries_reached}
+            end
```

Comment:
Just personal taste, but I would separate this into two separate functions:

```elixir
defp skip_failure?(true) do
  Logger.error(
    "#{__MODULE__} failed to handle event, skipping after #{max_retries} retries"
  )
  :skip
end
```

either the `false` case. But really, an if is ok too.

Reply:
Fair enough :+1: